lib/stages/VisRawReader.cpp

```diff
@@ -0,0 +1,112 @@
+    _prods = metadata_json["index_map"]["prod"].get<std::vector<prod_ctype>>();
+    _ev = metadata_json["index_map"]["ev"].get<std::vector<uint32_t>>();
+    if (metadata_json.at("index_map").find("stack") != metadata_json.at("index_map").end()) {
+        _stack = metadata_json.at("index_map").at("stack").get<std::vector<stack_ctype>>();
+        _rstack = metadata_json.at("reverse_map").at("stack").get<std::vector<rstack_ctype>>();
+        _num_stack = metadata_json.at("structure").at("num_stack").get<uint32_t>();
+    }
+
+    // Check metadata is the correct size
+    if (sizeof(VisMetadata) != metadata_size) {
+        std::string msg = fmt::format(fmt("Metadata in file {:s} is larger ({:d} bytes) than "
+                                          "VisMetadata ({:d} bytes)."),
+                                      filename, metadata_size, sizeof(VisMetadata));
+        throw std::runtime_error(msg);
+    }
+
+    // Register a state for the time axis if using comet, or register the replacement dataset ID if
+    // using
+    if (update_dataset_id) {
+
+        datasetManager& dm = datasetManager::instance();
+
+        if (!use_comet) {
+            // Add data specific states
+            if (!_stack.empty())
+                states.push_back(dm.create_state<stackState>(_num_stack, std::move(_rstack)).first);
+            states.push_back(dm.create_state<inputState>(_inputs).first);
+            states.push_back(dm.create_state<eigenvalueState>(_ev).first);
+            states.push_back(dm.create_state<prodState>(_prods).first);
+
+            // register it as root dataset
+            static_out_dset_id = dm.add_dataset(states);
+
+            WARN("Updating the dataset IDs without comet is not recommended "
+                 "as it will not preserve dataset ID changes.");
+        }
+    }
+}
+
+VisRawReader::~VisRawReader() {}
+
+void VisRawReader::create_empty_frame(frameID frame_id) {
+
+    // Create frame and set structural metadata
+    size_t num_vis = _stack.size() > 0 ? _stack.size() : _prods.size();
+
+    auto frame =
+        VisFrameView::create_frame_view(out_buf, frame_id, _inputs.size(), num_vis, _ev.size());
+
+    // Fill data with zeros
+    std::fill(frame.vis.begin(), frame.vis.end(), 0.0);
+    std::fill(frame.weight.begin(), frame.weight.end(), 0.0);
+    std::fill(frame.eval.begin(), frame.eval.end(), 0.0);
+    std::fill(frame.evec.begin(), frame.evec.end(), 0.0);
+    std::fill(frame.gain.begin(), frame.gain.end(), 0.0);
+    std::fill(frame.flags.begin(), frame.flags.end(), 0.0);
+    frame.erms = 0;
+
+    // Set non-structural metadata
+    frame.freq_id = 0;
+    frame.dataset_id = dset_id_t::null;
+    frame.time = std::make_tuple(0, timespec{0, 0});
+
+    // mark frame as empty by ensuring this is 0
+    frame.fpga_seq_length = 0;
+    frame.fpga_seq_total = 0;
+
+    DEBUG("VisRawReader: Reading empty frame: {:d}", frame_id);
+}
```

Comment:
Again, not really anything you've changed, but I think creating a zero'd out frame is generic enough I'd actually consider moving it into VisFrameView itself. Plus it can probably just be done with `memset` for more efficiency in there as VisFrameView can assume it knows the layout.

Also, I'm not even sure whether the frame needs to be zeroed out as you're just going to copy the actual data into it. Although I have a feeling that @tristpinsm might need it to send empty frames for missing data....

Reply:
that's right, in gossec these frames are eventually written to disk. If it's inconvenient to zero them here, you could always modify `visTranspose` to write zeroes when it detects an empty frame.

Reply:
Thanks for the info Tristan.